policy: Add listener priority

jrajahalme opened the pull request:
Note:

Add priority field to CNP listener reference, allowing for choosing which listener to redirect to when two listeners apply to the same identity and port. This is needed to allow one listener be used "by default" and another be used for specific destinations, for example.

The PR builds on these PRs that may affect possible backporting:
- #29023
- #29024
- #29025
- #29026
- #29162
- #29813
- #29815
- #30233 

```release-note
Cilium Network Policy can now redirect to different listeners on the same destination port depending on the destination.
```

jrajahalme committed:
endpoint: Add listener to proxyID

Add listener to the proxyID. This is needed so that different listeners
can be supported on the same port, for different destinations/sources.

The listener name also needs to be passed on via policy.MapStateEntry.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -172,6 +172,39 @@ func (e *Endpoint) writeHeaderfile(prefix string) error {
+
+func (p *proxyPolicy) GetListener() string {
+	return p.l4.GetListener()
+}
```
Some docblocks here would be nice. Doesn't have to be exhaustive, just a sentence or two.

jrajahalme replied:
This code is from https://github.com/cilium/cilium/pull/29023, I'll address this comment there.

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -225,9 +225,22 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
 		func(l4 *policy.L4Filter) (uint16, bool) {
 			var redirectPort uint16
+
+			proxyID, port := e.proxyID(l4)
```
These sorts of things make my head spin :-(. Is this the pod port, the peer port, or the proxy port?

jrajahalme replied:
renamed as `dstPort` and added this comment above:
```
+				// proxyID() returns also the destination port for the policy,
+				// which may be resolved from a named port
+				proxyID, dstPort := e.proxyID(l4)
```

squeed commented on pkg/policy/mapstate.go:
```diff
@@ -1193,6 +1210,8 @@ func (ms *mapState) AddVisibilityKeys(e PolicyOwner, redirectPort uint16, visMet
 				// 3. Change all L3/L4 ALLOW keys on matching port that do not
 				//    already redirect to redirect.
 				v.ProxyPort = redirectPort
+				v.priority = redirectPort
```
This is a bit confusing -- why is the port the same as the priority?

jrajahalme replied:
It's the other way around, by default the priority is the same as the redirect port number. This would normally be hidden in `NewMapStateEntry()` where this is also properly documented, but here MapStateEntries are modified in place. Should add a comment, though!

aanm replied:
Please do

jrajahalme replied:
Added comment!

squeed reviewed (commented):
Looks reasonable except for one question. I'm not marking as reviewed since I'm not an "assigned" reviewer.

pippolo84 commented on pkg/policy/proxyid.go:
```diff
@@ -12,6 +12,16 @@ import (
+	if ingress {
+		direction = "ingress"
+	}
+	return fmt.Sprintf("%s:%s:%d:%d", direction, protocol, port, proxyPort)
```
Not sure if this qualifies as a hot path, but it might be worth using a `strings.Builder` here (see https://github.com/cilium/cilium/issues/19571)

pippolo84 commented on pkg/policy/mapstate.go:
```diff
@@ -477,7 +483,7 @@ func (e *MapStateEntry) Merge(entry *MapStateEntry) {
 }
 
-// IsRedirectEntry returns true if e contains a redirect
+// IsRedirectEntry returns true if e contains a proxy ID
```
I find this comment a bit confusing. Shouldn't this be left untouched or changed to something like _if e contains a non zero proxy port_ ? Also, not related to this PR, but this method should be unexported.

jrajahalme replied:
You are correct, that is a weird change. Updated to:
```
// IsRedirectEntry returns true if the entry redirects to a proxy port
```

aanm reviewed (approved):
LGTM, but it will need more people from the policy side to look at the PR.

pippolo84 commented on pkg/policy/mapstate.go:
```diff
@@ -435,12 +443,20 @@ func (e *MapStateEntry) Merge(entry *MapStateEntry) {
+		if entry.IsRedirectEntry() {
+			// Lower number has higher priority
+			if !e.IsRedirectEntry() ||
+				(entry.ProxyPort != e.ProxyPort && entry.priority < e.priority) {
```
I'm not sure I'm correctly understanding the logic here. Shouldn't we overwrite `e.Listener` and `e.Priority` even when `entry.ProxyPort == e.ProxyPort`?

jrajahalme replied:
Each listener has a unique proxy port, so when the proxy ports are the same, the listeners should already have the same value. But this brings up the point that now that MapStateEntry contains the correct proxy port from the beginning (3rd commit), we might get away with without the listener name in the MapSteteEntry (added in the 2nd commit, as the proxy port was not initially valid). I'll check this possibility in a follow-up.

But you are absolutely correct about the priority, as it is not an intrinsic property of the listener itself, but of the reference to a listener. Different references to the same listener could have different priorities, which needs to be properly propagated. I'll fix this immediately. Turns out that the simplest form is what you were after, overwriting also values that may already be the same:
```
		// Proxy port takes precedence, but may be updated due to priority
		if entry.IsRedirectEntry() {
			// Lower number has higher priority, but non-redirects have 0 priority value
			if !e.IsRedirectEntry() || entry.priority < e.priority {
				e.ProxyPort = entry.ProxyPort
				e.Listener = entry.Listener
				e.priority = entry.priority
			}
		}
```
While adding unit testing for the above I found two additional problems:
1. I had missed updating the newish MapStateEntry.DeepEqual function when rebasing this to main (DeepEqual was added in #26344)
2. mapState.addKeyWithChanges() has a bug where DatapathEqual() is evaluated after the entries are already merged (so they are datapath-equal in all cases). This bug was introduced in #26331, which was backported to 1.14. Fixed in a separate commit, will note in PR description.

pippolo84 replied:
Thanks for the explanation :pray:

pippolo84 reviewed (commented):
Thanks for the PR! :rocket: 
The approach seems sound, I've left a suggestion and a couple of questions inline.

joestringer commented on pkg/policy/api/l4.go:
```diff
@@ -150,6 +150,19 @@ type Listener struct {
 	Name string `json:"name"`
+
+	// Priority is used to select the Listener if multiple rules would apply different listeners
+	// to a policy map entry. Lower numbers indicate higher priority. If left out, a random, but
```
I think there's one thing that this description doesn't super explicitly point out today, which could be clearer: The priority aspect is super clear but the 'control flow' is not as obvious to me. Maybe something like:

> The traffic will only be redirected to one listener.

jrajahalme replied:
Given the implementation dependent nature of this definition I heeded your overall concern and changed the API description to:
```
	// Priority for this Listener that is used when multiple rules would apply different
	// listeners to a policy map entry. Behavior of this is implementation dependent.
```

joestringer commented on pkg/policy/api/l4.go:
```diff
@@ -150,6 +150,19 @@ type Listener struct {
+	// consistently redirected to the same listener. If higher priority desired, a low unique
+	// number like 1, 2, or 3 should be explicitly specified here. If a lower than default
+	// priority is needed, then a unique number higher than 20000 should be explicitly
+	// specified. Numbers on the default range (10000-20000) are not allowed.
```
The 10-20K behaviour to me sounds like it's implementation specifics leaking out into the API without firmly deciding about the target use cases and how users will use these priorities. In recent k8s upstream APIs, they've added priorities and in that case just added a range of 0-1000 ([link](https://network-policy-api.sigs.k8s.io/reference/spec/#policy.networking.k8s.io%2fv1alpha1.AdminNetworkPolicy)). I'm somewhat inclined to think that we don't need this level of complication expressed in the API.

If someone has listener policies today, then the ordering is arbitrary. If they upgrade and they want a particular listener to be used with priority vs another, then I think it's reasonable to expect them to define the priority for both listeners.

I'm also not quite sure what happens if someone starts to use a significant share of the 65K listeners, will that have bad bounding of the worst case behaviour? Limiting to 1K seems like it provides plentiful opportunities for use cases while reducing the scope for abuse. (I might argue even 100 is enough, easier to start with a smaller value and increase it later rather than start with a higher value and start to invalidate user policies later).

jrajahalme replied:
Yes, way too implementation specific for the API spec, agreed. We currently have no specific limit to the number of listeners, but I think there is a limit to the open file descriptors Envoy can hold imposed by the OS.

joestringer commented on pkg/policy/api/l4.go:
```diff
@@ -150,6 +150,19 @@ type Listener struct {
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65535
+	// +kubebuilder:validation:Optional
+	Priority uint16 `json:"priority"`
```
Let's say the cluster administrator creates a listener policy in a CCNP for a particular port, then a namespace user creates another listener policy in CNP for a port, I guess that means that the namespace user can override/preempt the clusterwide configuration? I'm trying to think a little bit outside the box about how the semantics of this field may apply in conjunction with other policy primitives we have.

One might argue that if the cluster administrator wants to ensure that a particular listener applies on a port, then they could do so with priority 1, and if they wanted to provide a way for it to be overridden, they could select a more deprioritized value. But then, what happens if both policies have priority 1? I guess we fall back to random / implementation-defined behaviour?

jrajahalme replied:
Interesting aspect that warrants further thought. Currently, any deny policies in a CCNP get precedence over any allows in a CNP (like they do in a single CNP). Maybe listener references in CCNPs should get precedence over those in CNPs? This also points to the direction that we can't nail down the API spec for this at this moment, but must leave it open for improvement.

joestringer replied:
Do we just implement this by parsing the priorities and adding an offset to them for CNP?

Ie if there's a CCNP with a priority, it gets the priority specified by the user, maybe bounded 1-100 for example. If there's a CNP, after Cilium parses the priority it adds the max value (100) so the CNP gets priority 101 for user input `priority: 1` etc.

Or for now we just leave it undefined until we have a stronger set of user stories to back up the API feature.

tommyp1ckles replied:
Would that impose a limit of max 99 CNPs with listener priority? Not that this limit would necessarily be unreasonable, but we'd probably have to express that in the API somehow then.

joestringer replied:
For that proposal it would not limit the number of CNPs that someone can create. We would still need to define that if there are two CNPs selecting the same traffic with the same listener priority then the behaviour is undefined. All it would do is impose an ordering between CCNP vs. CNP.

joestringer reviewed (commented):
A few comments on the API, I didn't review the rest (the commit messages seemed like they logically followed and I see other reviewers looking in further detail.)

Some part of me wonders whether priority should be made more of an opaque API since it's based on the behaviour of the CEC resources which are non-stable API. Ie just define the priority as "implementation defined". Given that the changes to network policy are stable API, it's a bit weird to define that the new field would have specific behaviour if the underlying mechanism is completely subject to change.

tommyp1ckles commented on pkg/endpoint/bpf.go:
```diff
@@ -265,8 +265,9 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
+				statsKey := e.proxyStatsKey(l4, redirectPort)
 				e.proxyStatisticsMutex.Lock()
-				proxyStats := e.getProxyStatisticsLocked(proxyID, string(l4.L7Parser), uint16(l4.Port), l4.Ingress)
+				proxyStats := e.getProxyStatisticsLocked(statsKey, string(l4.L7Parser), uint16(l4.Port), l4.Ingress)
```
proxyStatsKey can be an empty string if port is 0, should we check against this case?

jrajahalme replied:
Will address by removing this helper as the named port is nowadays actually already resolved when proxy ID is formatted (before the use here)

tommyp1ckles commented on pkg/endpoint/redirect_test.go:
```diff
@@ -483,3 +484,184 @@ func (s *RedirectSuite) TestRedirectWithDeny(c *check.C) {
+	}
+	do.repo.AddList(rules)
+
+	//logging.SetLogLevelToDebug()
```
leftover commented code?

joestringer commented on pkg/policy/api/l4.go:
```diff
@@ -150,6 +150,14 @@ type Listener struct {
+	// listeners to a policy map entry. Behavior of this is implementation dependent.
+	//
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65535
```
Even if it's implementation dependent, I think that starting with a smaller number provides better bounding of worst case behaviour. I'm a little concerned about having an API we can't express because of local fd limits.

```suggestion
	// +kubebuilder:validation:Maximum=100
```

jrajahalme replied:
Sorry, earlier thought the limited number referred to the number of listeners in general! Totally makes sense to limit the range here to begin with.

jrajahalme replied:
Done!

pippolo84 commented on pkg/policy/rule.go:
```diff
@@ -162,6 +153,20 @@ func mergePortProto(ctx *SearchContext, existingFilter, filterToMerge *L4Filter,
+			if newL7Rules.Listener != l7Rules.Listener && newL7Rules.Priority < l7Rules.Priority {
+				l7Rules.Listener = newL7Rules.Listener
+			}
+
```
What about using a helper like the following:

```
func mergeListenerReference(ctx *SearchContext, l7Rules, newL7Rules *PerSelectorPolicy) error {
	if newL7Rules.Listener == "" {
		return nil
	}

	if l7Rules.Listener == "" {
		l7Rules.Listener = newL7Rules.Listener
		return nil
	}

	if l7Rules.Listener == newL7Rules.Listener || newL7Rules.Priority > l7Rules.Priority {
		return nil
	}

	if newL7Rules.Priority < l7Rules.Priority {
		l7Rules.Listener = newL7Rules.Listener
		return nil
	}

	// conflict: newL7Rules.Priority == l7Rules.Priority
	ctx.PolicyTrace("   Merge conflict: mismatching CiliumEnvoyConfig listeners %v/%v with the same priority %d\n", newL7Rules.Listener, l7Rules.Listener, l7Rules.Priority)
	return fmt.Errorf("cannot merge conflicting CiliumEnvoyConfig Listeners (%v/%v) with the same priority (%d)", newL7Rules.Listener, l7Rules.Listener, l7Rules.Priority)
}
```

Also, I think it'd be better to add a test for this.
Finally, why the priority is not overwritten when the newL7Rules Listener is written over the existing one?

jrajahalme replied:
Added a helper and tests!

jrajahalme replied:
> Finally, why the priority is not overwritten when the newL7Rules Listener is written over the existing one?

That was a bug that is now fixed and also covered by the new unit tests!

pippolo84 reviewed (changes requested):
Thanks for the further iteration!
One last comment about `mergePortProto`, apart from this it seems good to me.

tommyp1ckles commented on pkg/endpoint/bpf.go:
```diff
@@ -212,76 +218,79 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
+				// to 0.
+				if !e.hasSidecarProxy || l4.L7Parser != policy.ParserTypeHTTP {
+
+					pp := e.newProxyPolicy(l4, v, dstPort)
```
Is there any potential downsides with creating separate proxy redirects per selector policy, even if they're the same listener?

jrajahalme replied:
We use `desiredRedirects[proxyID]` to skip creating/updating the same redirect multiple times a few lines above:
```
				// desiredRedirects starts out empty, so we can use it check
				// if the redirect has already been updated on this round.
				if desiredRedirects[proxyID] {
					continue
				}
```

tommyp1ckles commented on pkg/endpoint/bpf.go:
```diff
@@ -212,76 +218,79 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
+					if err != nil {
+						// Skip redirects that can not be created or updated.  This
+						// can happen when a listener is missing, for example when
+						// restarting and k8s delivers the CNP before the related
```
What is the implication of this, would this cause dropped flows on that port until the CEC is processed?

jrajahalme replied:
Yes, packets hitting policies with listener reference can get dropped due to listener not existing yet. Any CEC/CCEC ingestion causes the endpoint policies to be regenerated so that any drops due to this are only transient.

tommyp1ckles commented on pkg/endpoint/bpf.go:
```diff
@@ -212,76 +218,79 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
+				}
+				if _, found := e.realizedRedirects[proxyID]; !found {
+					revertStack.Push(func() error {
+						delete(e.realizedRedirects, proxyID)
```
The corresponding "revertFunc" for this can fail with an error, do we still want to remove this from the realized state if that fails?

tommyp1ckles replied:
I guess this is just reshuffling of existing code, so this would be keeping the existing behavior.

jrajahalme replied:
In practice we never fail revert functions, nor check their error return. This is an area for clean-up, IMO the revert function should not return an error.

tommyp1ckles commented on pkg/endpoint/bpf.go:
```diff
@@ -212,76 +218,79 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
-			return redirectPort, true
+			// Any map updates when a new policy has not been calculated are taken care
+			// by incremental map updates, so return 'false' in that case.
+			return e.desiredPolicy != e.realizedPolicy
```
Bit of a tangent, but I've been a bit confused by the meaning of these comparisons in Endpoint code,  from what I can the desired/realized policy both reference the same underlying so this should always return true.

jrajahalme replied:
They are the same pointer in steady state, but during a policy reconfiguration they are different. The pattern is that `e.desiredPolicy` is set to a newly computed policy, and then when it has been fully realized, `e.updateRealizedState()` is called to update `e.realizedPolicy` to point to the same policy.

jrajahalme replied:
It is possible that `addNewRedirectsFromDesiredPolicy()` is only called after the desired policy has been set, so the pointer would always be different here, will investigate.

jrajahalme replied:
`e.desiredPolicy == e.realizedPolicy` when endpoint regeneration is triggered and the current policy has already been computed for this endpoint. It seems harmless to add the redirect entries to the policy map also in this case (even though it is likely a no-op). I added a note on this to `runPrecompilationSteps()`: 
```
 	// Set the computed policy as the "incoming" policy. This can fail if
 	// the endpoint's security identity changed during or after policy calculation.
+	// Note that the incoming policy can be the same as the previous policy in cases
+	// where an unnecessary policy computation was skipped. In that case
+	// e.desiredPolicy == e.realizedPolicy also after this call.
+	// TODO: Figure out how much of the following logic would be safe to skip if there
+	// is no new incoming policy.
 	if err := e.setDesiredPolicy(policyResult); err != nil {
 		return false, err
 	}
```

jrajahalme replied:
Turns out incremental updates will start being bypassed if a redirect in a `desiredPolicy` mapstate is updated when it is in fact the same as `realizedPolicy`, I think what happens is that the datapath bpf map update is skipped due to the entry already existing in the `realizedPolicy`. So I had to keep the pointer comparison logic here.

tommyp1ckles replied:
I was looking through this code a bit a while back trying to figure out what situations the desiredPolicy would be recreated. At the time it appeared to me that having the two separate maps was not actually meaningful since there was never a case where `e.desiredPolicy != e.realizedPolicy`

Even in the `e.setDesiredPolicy(policyResult)` case, if I look at where the `endpointPolicy` that the desiredPolicy is being reassigned to comes from, its just the same reference of e.desiredPolicy:

```go
	result := &policyGenerateResult{
		selectorPolicy:   e.selectorPolicy,
		endpointPolicy:   e.desiredPolicy,
		identityRevision: e.identityRevision,
	}
```

It seemed like desiredPolicy is only ever assigned to a new map when being constructured for newEndpoint/restoreEndpoint.

jrajahalme replied:
This is where it is updated once a new policy has been calculated:
```
	// Consume converts a SelectorPolicy in to an EndpointPolicy
	result.endpointPolicy = result.selectorPolicy.Consume(e)
```

jrajahalme replied:
And then:
```
	// Set the revision of this endpoint to the current revision of the policy
	// repository.
	e.setNextPolicyRevision(res.policyRevision)
	e.selectorPolicy = res.selectorPolicy
	e.desiredPolicy = res.endpointPolicy
```

tommyp1ckles replied:
ah I see now, distill allocates a new MapState which gets repopulated.

pippolo84 reviewed (approved):
LGTM, thanks! :100:

tommyp1ckles commented on pkg/endpoint/bpf.go:
```diff
@@ -210,46 +216,55 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
-
+		func(l4 *policy.L4Filter) bool {
+			for _, v := range l4.PerSelectorPolicies {
+				if v == nil || !v.IsRedirect() {
```
When is v expected to be nil here?  If it's not expected, and this is just guarding against that, perhaps we should log something.

jrajahalme replied:
`v` is expected to be `nil` for an allow selector that has no auth or L7 rules.

tommyp1ckles commented on pkg/envoy/accesslog_server.go:
```diff
@@ -162,7 +162,7 @@ func (s *AccessLogServer) handleConn(ctx context.Context, conn *net.UnixConn) {
 				port = r.SourceEndpoint.Port
 			}
-			localEndpoint.UpdateProxyStatistics("envoy", "TCP", port, ingress, request, r.Verdict)
+			localEndpoint.UpdateProxyStatistics("envoy", "TCP", port, uint16(pblog.ProxyId), ingress, request, r.Verdict)
```
Question, is there a case where ProxyID might not be set in the Envoy log message, causing a collision on the port 0 visibility stat?

jrajahalme replied:
Good question, as I have managed to not set the `ProxyId` value in the Listener config we send to Envoy. It is a new field in the config of `cilium.bpf_metadata` listener filter, will add it to this commit in an update.

And you are right, with the `ProxyId` zero the stats for different listeners on the same destination port would be collapsed together.

jrajahalme replied:
Fixed, please re-review the (updated) commit "endpoint: Add proxy port to proxy stats key", which now also sets the `ProxyId` in `pkg/envoy/xds_server.go`

jrajahalme committed:
endpoint: Make realizedRedirects lockless

Endpoint's realizedRedirects is used to look up a proxy port for a
redirect in during policy map updates. Make access to it lockless by
storing an atomic pointer to the map, and considering the stored map
immutable.

The set of realized redirects initially starts empty, and all required
(desired) redirects are added to it. After that the unwanted redirects
are removed by comparing the old realized redirects and the new desired
redirects. After this the desired redirects becomes the new realized
redirects.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
endpoint: Add proxy port to proxy stats key

Proxy stats contains the destination port of redirected traffic. When a
single port can be redirected to multiple listeners, depending on the
destination (or source), their stats entries need to be kept
separate. One way of doing this is to add the proxy port to the proxy
stats key.

Proxy port is wired through the ProxyId field in the cilium.bpf_metadata
filter config, and will be carried over to the access log messages from
there.

Proxy stats are endpoint specific, so the endpoint id need not be in
proxy stats key.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -748,6 +741,13 @@ func (e *Endpoint) runPreCompilationSteps(regenContext *regenerationContext, rul
+	// TODO: Figure out how much of the following logic would be safe to skip if there
+	// is no new incoming policy. Policy update can be triggered for changes in named ports
+	// and Envoy listeners, so it seems policy maps and Envoy network policy will need to
+	// be recomputed regardless.
```
Do you have plans to follow up on this TODO? Typically TODOs that are added to a codebase just get ignored and become stale, so we typically avoid adding them into the tree in the first place.

There's already some degree of skipping particularly for BPF compilation/loading if the regeneration wasn't triggered based on the need for that (via `datapathRegenCtxt`), otherwise in general this seems to just be the regular configuration path. So it seems like there's some potential for early exit here, but yeah I do think that should probably be investigated subsequently as part of a dedicated PR.

jrajahalme replied:
Agree, TODO's in source tend to be forgotten, removed. A dedicated PR should be brought forward once someone has a concrete idea what to do about this if anything.

joestringer commented on pkg/k8s/apis/cilium.io/register.go:
```diff
@@ -15,5 +15,5 @@ const (
 	// Maintainers: Run ./Documentation/check-crd-compat-table.sh for each release
 	// Developers: Bump patch for each change in the CRD schema.
-	CustomResourceDefinitionSchemaVersion = "1.26.12"
+	CustomResourceDefinitionSchemaVersion = "1.26.13"
```
Ah looks like we've missed updating this. Cilium v1.15 will use 1.28.x and Cilium v1.16 (main) will use 1.29.x.
```suggestion
	CustomResourceDefinitionSchemaVersion = "1.29.0"
```

joestringer replied:
(I already bumped the other stable branches last year, only `main` is out of sync)

jrajahalme replied:
Saw this when rebasing to main :-)

joestringer commented on pkg/policy/rule.go:
```diff
@@ -162,6 +195,12 @@ func mergePortProto(ctx *SearchContext, existingFilter, filterToMerge *L4Filter,
 			l7Rules.isRedirect = l7Rules.isRedirect || newL7Rules.isRedirect
 
+			// Merge listener reference
+			if err := l7Rules.mergeListenerReference(newL7Rules); err != nil {
```
I think that this is resolving the case where you have two rules selecting the same traffic with different priorities + L7 configuration, eg:

- Traffic to app=foo on port 80/TCP should use listener A with prio 1
- Traffic to app=foo on port 80/TCP should use listener B with prio 2

I think there's another case, which admittedly is probably covered "accurately" by the current API since it's deliberately opaque but consider:

- Traffic to app=foo on port 80/TCP should use listener A with prio 1
- Traffic to class=empire on port 80/TCP should use listener B with prio 1

Given that these are acting on different selectors, I expect them to be encoded in different `*PerSelectorPolicy` objects. However, it's possible for each of these rules to select the same identity if that identity includes both `app=foo`,`class=empire`. In that case, the priority conflict would make the rules impossible to evaluate.

I don't currently see code handling that case. (sidenote, how do we test this?)

jrajahalme replied:
Thanks for reminding about this. Currently equal priority will revert back to the prior behavior where the result depends on the (random) order of generating the mapstate entries. I'll need to add the use of the proxy port number as a tie-breaker, so it will be deterministic at least.

jrajahalme replied:
Added a new unit test `TestRedirectWithEqualPriority()` to test for the case where priority is explicitly given as the same value for two different listeners in policy rules for different selectors selecting the same remote ID.

joestringer reviewed (commented):
API LGTM now. I also just took a fresh overview look and identified a couple of minor areas you may want to follow up on / adjust.

mhofstetter reviewed (approved):
`envoy` :heavy_check_mark: 

Thanks Jarno

mhofstetter commented on pkg/envoy/xds_server.go:
```diff
@@ -805,12 +805,13 @@ func (s *xdsServer) deleteSecret(name string, wg *completion.WaitGroup, callback
 		UseOriginalSourceAddress: useOriginalSourceAddr,
 		BpfRoot:                  bpf.BPFFSRoot(),
 		IsL7Lb:                   l7lb,
+		ProxyId:                  uint32(proxyPort),
```
nit: maybe add a comment why we're setting the proxy port on the field `ProxyId`. This could look like an error at first glance.

we could do the same in the `accessloger_server.go` were we're doing the same in reverse (interpreting the proxyid from the access log entry as proxyPort) when passing it on to the `EndpointUpdater`.

joestringer commented on pkg/endpoint/redirect_test.go:
```diff
@@ -478,6 +482,315 @@ func (s *RedirectSuite) TestRedirectWithDeny(c *check.C) {
+	}
+	do.repo.GetSelectorCache().SetLocalIdentityNotifier(testidentity.NewDummyIdentityNotifier())
+	identitymanager.Subscribe(do.repo)
+	defer identitymanager.RemoveAll()
```
(There's a bit of code here that seems like boilerplate across multiple tests, does it differ? Is there an opportunity for a helper here or is it awkward to do so?)

jrajahalme replied:
Good call. Added new `SetUpTest()` `TearDownTest()` and helpers for Endpoint and policy rules to remove this and more of the boilerplate.

joestringer commented on pkg/endpoint/redirect_test.go:
```diff
@@ -478,6 +482,315 @@ func (s *RedirectSuite) TestRedirectWithDeny(c *check.C) {
+		ruleL4AllowPort80.WithEndpointSelector(selectBar_),
+		ruleL4L7AllowListener2Priority1.WithEndpointSelector(selectBar_),
+	}
+	do.repo.AddList(rules)
```
As I understand, this test was written in response to questions about this case:

- Traffic to app=foo on port 80/TCP should use listener A with prio 1
- Traffic to class=empire on port 80/TCP should use listener B with prio 1
- Each of these rules to select the same identity if that identity includes both app=foo,class=empire

(cf. https://github.com/cilium/cilium/pull/28555#discussion_r1445504871)

These rules seem to cover the "same priority" aspect of testing, using the same selector. Does this test also cover the "disjoint selector" aspect? The reason I ask is that I can understand and saw the code that handles having a single selector which gets merged with another selector due to the selector+port collision, but I didn't understand whether it covers the case where the selectors end up distinct inside `l4.PerSelectorPolicies`.

jrajahalme replied:
`WithEndpointSelector()` here specifies which endpoints the rule applies to:
```
type Rule struct {
	// EndpointSelector selects all endpoints which should be subject to
	// this rule. EndpointSelector and NodeSelector cannot be both empty and
	// are mutually exclusive.
	//
	// +kubebuilder:validation:OneOf
	EndpointSelector EndpointSelector `json:"endpointSelector,omitempty"`
```

The rules themselves have different `FromEndpoints` selectors as intended:
- `ruleL4L7AllowListener1Priority1`: `FromEndpoints: []api.EndpointSelector{selectFoo_}`
- `ruleL4L7AllowListener2Priority1`: `FromEndpoints: []api.EndpointSelector{selectRed_}`

joestringer replied:
Ah right. One of these days we should come up with canonical terminology for the target policy enforcement point vs. the peer selector. Makes sense, thanks!

jrajahalme committed:
policy: Move Listener from L4Filter to PerSelectorPolicy

Allow different selectors on the same L4Filter use different Envoy
Listeners. This relaxes the policy import (L4Filter merge) logic by only
failing out if there is a Listener conflict on the same cached selector.

This change is needed to allow different Envoy Listeners to be applied on
traffic on the same port, depending on the destination (for an egress
policy). Consequently, we must handle conflicting proxy ports on the same
MapState key, originating from different selectors selecting the same
remote identity. We do this with a new optional Listener priority value.

Listener priority, if not specified, or for redirects for which an
explicit listener name is not given, defaults to the value of the proxy
port itself. This serves as a tie-breaking rule so that the redirection
is deterministic also in cases where a policy with a listener reference
and a CNP L7 policy on a different selectors that then happen to select
the same identities.

The proxy port value is also used as a tie-breaker when the same identity
is selected by two different selectors on different rules that specify
different listeners but with the same priority. While this is an
arbitrary choice, it is better than allowing the selected listener vary
depending on rule insertion order, or the random Go map iteration order
when generating the map state.

By convention proxy port values are between 10000-20000, so defining any
(allowed) priority value gives precedence to that listener reference
against listener references without an explicit priority.

Use MapState.Diff to report the difference between the obtained and
expected MapState on test failures.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Look up proxy port when creating mapstate entries

Populate mapstate entries with the actual proxy redirection port number
when creating MapState entries. New redirects may not be realized at the
point when new identities related to them are first accumulated. Thus the
accumulated MapChanges may still use in invalid proxy port internally,
but that is resolved before exposing the accumulated mapstate to the
endpoint package via ConsumeMapChanges.

Endpoint.realizedRedirects map now holds zero valued redirect ports for
Istio sidecars, so that the lookup can be made without taking endpoint's
mutex (without using Endpoint.hasSidecarProxy). Zero valued redirects are
not created or removed from the proxy package.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
